File: typescript/tests/memory_sync_service.test.ts
// memory_sync_service.test.ts
//
// Verifies the CircleAI.Sync port (MemorySyncService): push builds a broadcast
// SyncDelta with the local device as source and empty target, and the receive
// loop deserialises episodic deltas into the local store while skipping its own
// echoes and non-episodic domains. Also covers the JSON episodic codec and the
// InMemoryGoalStore added to the memory stores module.

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  MemorySyncService,
  JsonEpisodicDeltaCodec,
  SyncDeliveryMode,
  SyncDomainKeys,
  type ISyncChannel,
  type SyncDelta,
} from '../src/sync/index';
import { InMemoryEpisodicStore, InMemoryGoalStore } from '../src/memory/stores';
import { Goal, GoalStatus, GoalPriority, type EpisodicMemoryEntry } from '../src/memory/index';

// ─────────────────────────────────────────────────────────────────────────────
// A controllable in-process ISyncChannel: pushes are recorded; the receive
// stream yields whatever the test enqueues, then ends.
// ─────────────────────────────────────────────────────────────────────────────

class FakeSyncChannel implements ISyncChannel {
  readonly pushed: SyncDelta[] = [];
  private readonly inbound: SyncDelta[];

  constructor(inbound: SyncDelta[] = []) {
    this.inbound = inbound;
  }

  pushDeltaAsync(delta: SyncDelta): Promise<void> {
    this.pushed.push(delta);
    return Promise.resolve();
  }

  async *receiveDeltasAsync(_ownerId: string): AsyncGenerator<SyncDelta> {
    for (const d of this.inbound) {
      yield d;
    }
  }

  getLastSequenceAsync(_ownerId: string, _domainKey: string): Promise<number> {
    return Promise.resolve(0);
  }
}

function episodicEntry(id: string, text: string): EpisodicMemoryEntry {
  return {
    id,
    recordedAtUtc: new Date('2026-04-04T12:00:00Z'),
    userText: text,
    assistantText: `re: ${text}`,
    appContext: 'test',
    tags: { topic: 'demo' },
  };
}

function delta(over: Partial<SyncDelta> & Pick<SyncDelta, 'ownerId' | 'sourceDeviceId' | 'domainKey' | 'payload'>): SyncDelta {
  return {
    ownerId: over.ownerId,
    sourceDeviceId: over.sourceDeviceId,
    targetDeviceId: over.targetDeviceId ?? '',
    domainKey: over.domainKey,
    payload: over.payload,
    sequence: over.sequence ?? 1,
    deliveryMode: over.deliveryMode ?? SyncDeliveryMode.GUARANTEED,
    ttlMs: over.ttlMs,
    createdAt: over.createdAt ?? new Date('2026-04-04T12:00:00Z'),
  };
}

// ─────────────────────────────────────────────────────────────────────────────
// JsonEpisodicDeltaCodec
// ─────────────────────────────────────────────────────────────────────────────

describe('JsonEpisodicDeltaCodec', () => {
  it('round-trips an episodic entry', () => {
    const codec = new JsonEpisodicDeltaCodec();
    const original = episodicEntry('e1', 'hello');
    const decoded = codec.decode(codec.encode(original));
    assert.ok(decoded);
    assert.equal(decoded!.id, 'e1');
    assert.equal(decoded!.userText, 'hello');
    assert.equal(decoded!.assistantText, 're: hello');
    assert.equal(decoded!.appContext, 'test');
    assert.equal(decoded!.tags?.topic, 'demo');
    assert.equal(decoded!.recordedAtUtc.toISOString(), '2026-04-04T12:00:00.000Z');
  });

  it('returns null for garbage bytes', () => {
    const codec = new JsonEpisodicDeltaCodec();
    assert.equal(codec.decode(new TextEncoder().encode('not json')), null);
    assert.equal(codec.decode(new TextEncoder().encode('{"no":"id"}')), null);
  });
});

// ─────────────────────────────────────────────────────────────────────────────
// MemorySyncService — push
// ─────────────────────────────────────────────────────────────────────────────

describe('MemorySyncService — pushMemoryDelta', () => {
  it('builds a broadcast delta stamped with the local device id', async () => {
    const channel = new FakeSyncChannel();
    const store = new InMemoryEpisodicStore();
    const fixedNow = new Date('2026-05-05T00:00:00Z');
    const svc = new MemorySyncService(channel, store, 'device-local', new JsonEpisodicDeltaCodec(), () => fixedNow);

    const payload = new Uint8Array([1, 2, 3]);
    await svc.pushMemoryDeltaAsync('owner-1', SyncDomainKeys.EPISODIC_MEMORY, payload, SyncDeliveryMode.URGENT);

    assert.equal(channel.pushed.length, 1);
    const d = channel.pushed[0];
    assert.equal(d.ownerId, 'owner-1');
    assert.equal(d.sourceDeviceId, 'device-local');
    assert.equal(d.targetDeviceId, '', 'broadcast → empty target');
    assert.equal(d.domainKey, SyncDomainKeys.EPISODIC_MEMORY);
    assert.deepEqual([...d.payload], [1, 2, 3]);
    assert.equal(d.deliveryMode, SyncDeliveryMode.URGENT);
    assert.equal(d.sequence, fixedNow.getTime());
    assert.equal(d.createdAt.getTime(), fixedNow.getTime());
  });

  it('defaults the delivery mode to Guaranteed', async () => {
    const channel = new FakeSyncChannel();
    const svc = new MemorySyncService(channel, new InMemoryEpisodicStore(), 'dev');
    await svc.pushMemoryDeltaAsync('o', SyncDomainKeys.PERSONA, new Uint8Array([0]));
    assert.equal(channel.pushed[0].deliveryMode, SyncDeliveryMode.GUARANTEED);
  });
});

// ─────────────────────────────────────────────────────────────────────────────
// MemorySyncService — receive
// ─────────────────────────────────────────────────────────────────────────────

describe('MemorySyncService — receive loop', () => {
  it('applies inbound episodic deltas from other devices into the store', async () => {
    const codec = new JsonEpisodicDeltaCodec();
    const inbound = [
      delta({
        ownerId: 'owner-1',
        sourceDeviceId: 'device-remote',
        domainKey: SyncDomainKeys.EPISODIC_MEMORY,
        payload: codec.encode(episodicEntry('remote-1', 'from the phone')),
      }),
    ];
    const channel = new FakeSyncChannel(inbound);
    const store = new InMemoryEpisodicStore();
    const svc = new MemorySyncService(channel, store, 'device-local', codec);

    await svc.startReceivingAsync('owner-1');
    // Let the fire-and-forget receive loop drain the finite inbound stream.
    for (let i = 0; i < 20; i++) await Promise.resolve();
    await svc.stopReceivingAsync();

    assert.equal(await store.countAsync(), 1);
    const recent = await store.getRecentAsync(10);
    assert.equal(recent[0].id, 'remote-1');
    assert.equal(recent[0].userText, 'from the phone');
  });

  it('skips its own echoes (matching source device id)', async () => {
    const codec = new JsonEpisodicDeltaCodec();
    const inbound = [
      delta({
        ownerId: 'owner-1',
        sourceDeviceId: 'device-local', // our own echo
        domainKey: SyncDomainKeys.EPISODIC_MEMORY,
        payload: codec.encode(episodicEntry('echo', 'mine')),
      }),
    ];
    const svc = new MemorySyncService(new FakeSyncChannel(inbound), new InMemoryEpisodicStore(), 'device-local', codec);
    const store = (svc as unknown as { store: InMemoryEpisodicStore }).store;

    await svc.startReceivingAsync('owner-1');
    for (let i = 0; i < 20; i++) await Promise.resolve();
    await svc.stopReceivingAsync();

    assert.equal(await store.countAsync(), 0, 'own echo must not be applied');
  });

  it('ignores non-episodic domains', async () => {
    const codec = new JsonEpisodicDeltaCodec();
    const inbound = [
      delta({
        ownerId: 'owner-1',
        sourceDeviceId: 'device-remote',
        domainKey: SyncDomainKeys.PERSONA, // not episodic
        payload: codec.encode(episodicEntry('persona', 'nope')),
      }),
    ];
    const store = new InMemoryEpisodicStore();
    const svc = new MemorySyncService(new FakeSyncChannel(inbound), store, 'device-local', codec);

    await svc.startReceivingAsync('owner-1');
    for (let i = 0; i < 20; i++) await Promise.resolve();
    await svc.stopReceivingAsync();

    assert.equal(await store.countAsync(), 0, 'non-episodic domain must be ignored by the episodic handler');
  });
});

// ─────────────────────────────────────────────────────────────────────────────
// InMemoryGoalStore
// ─────────────────────────────────────────────────────────────────────────────

function goal(id: string, userId: string, status: GoalStatus): Goal {
  const g = new Goal();
  g.id = id;
  g.userId = userId;
  g.title = `Goal ${id}`;
  g.status = status;
  g.priority = GoalPriority.Normal;
  g.createdUtc = new Date('2026-01-01T00:00:00Z');
  return g;
}

describe('InMemoryGoalStore', () => {
  it('upserts, gets, lists, filters active, and deletes', async () => {
    const store = new InMemoryGoalStore();
    await store.upsertAsync(goal('g1', 'u1', GoalStatus.Active));
    await store.upsertAsync(goal('g2', 'u1', GoalStatus.Completed));
    await store.upsertAsync(goal('g3', 'u2', GoalStatus.Active));

    assert.equal((await store.getAsync('g1'))?.id, 'g1');
    assert.equal(await store.getAsync('missing'), null);

    const u1 = await store.listAsync('u1');
    assert.deepEqual(u1.map((g) => g.id).sort(), ['g1', 'g2']);

    const activeU1 = await store.getActiveAsync('u1');
    assert.deepEqual(activeU1.map((g) => g.id), ['g1']);

    await store.deleteAsync('g1');
    assert.equal(await store.getAsync('g1'), null);
    assert.deepEqual((await store.listAsync('u1')).map((g) => g.id), ['g2']);
  });

  it('upsert replaces an existing goal by id and returns it', async () => {
    const store = new InMemoryGoalStore();
    await store.upsertAsync(goal('g1', 'u1', GoalStatus.Active));
    const updated = goal('g1', 'u1', GoalStatus.Abandoned);
    const returned = await store.upsertAsync(updated);
    assert.equal(returned.status, GoalStatus.Abandoned);
    assert.equal((await store.getAsync('g1'))?.status, GoalStatus.Abandoned);
    assert.equal((await store.listAsync('u1')).length, 1, 'upsert must not duplicate');
  });

  it('rejects blank ids / user ids', async () => {
    const store = new InMemoryGoalStore();
    await assert.rejects(() => store.getAsync(''), /id required/);
    await assert.rejects(() => store.listAsync('  '), /userId required/);
    await assert.rejects(() => store.getActiveAsync(''), /userId required/);
  });
});

File: rust/tests/memory_sync_service_test.rs
//! memory_sync_service_test.rs
//!
//! Verifies the `CircleAI.Sync` port: the [`MemorySyncService`] push/receive
//! orchestrator (push builds a broadcast SyncDelta; receive applies episodic
//! deltas, skips own echoes and other domains, and is gated by start/stop), the
//! [`SyncReconciliation`] version-vector helpers, and the in-memory sync
//! [`IGoalStore`] (`InMemoryGoalStore`).

use std::collections::HashMap;
use std::convert::Infallible;
use std::sync::Mutex;

use chrono::{TimeZone, Utc};
use circle_ai::memory::stores::{
    EpisodicMemoryEntry, IEpisodicMemoryStore, IGoalStore, InMemoryGoalStore,
};
use circle_ai::memory::{Goal, GoalPriority, GoalStatus};
use circle_ai::sync::{
    ISyncChannel, SyncDeliveryMode, SyncDelta, SyncDomainKeys, SyncReconciliation, VersionVector,
};
use circle_ai::sync_service::{IMemorySyncService, MemorySyncService};

// ── fakes ───────────────────────────────────────────────────────────────────

/// A channel that yields a scripted inbound queue (used by the receive tests;
/// pushes are ignored here — the push path is covered via `FakeChannelHandle`).
#[derive(Default)]
struct FakeChannel {
    inbound: Mutex<Vec<SyncDelta>>,
}

impl FakeChannel {
    fn with_inbound(inbound: Vec<SyncDelta>) -> Self {
        Self {
            inbound: Mutex::new(inbound),
        }
    }
}

impl ISyncChannel for FakeChannel {
    type Error = Infallible;

    fn push_delta(&mut self, _delta: &SyncDelta) -> Result<(), Self::Error> {
        Ok(())
    }

    fn receive_deltas(
        &self,
        _owner_id: &str,
    ) -> Result<Box<dyn Iterator<Item = Result<SyncDelta, Self::Error>>>, Self::Error> {
        // Drain the scripted inbound queue.
        let drained: Vec<SyncDelta> = std::mem::take(&mut *self.inbound.lock().unwrap());
        Ok(Box::new(drained.into_iter().map(Ok)))
    }

    fn get_last_sequence(&self, _owner_id: &str, _domain_key: &str) -> Result<i64, Self::Error> {
        Ok(0)
    }
}

/// A minimal in-memory episodic store implementing the sync trait.
#[derive(Default)]
struct VecEpisodicStore {
    entries: Mutex<Vec<EpisodicMemoryEntry>>,
}

impl IEpisodicMemoryStore for VecEpisodicStore {
    type Error = Infallible;

    fn add(&mut self, entry: EpisodicMemoryEntry) -> Result<(), Self::Error> {
        self.entries.lock().unwrap().push(entry);
        Ok(())
    }
    fn search(
        &self,
        _query_embedding: Option<&[f32]>,
        _top_k: usize,
    ) -> Result<Vec<EpisodicMemoryEntry>, Self::Error> {
        Ok(self.entries.lock().unwrap().clone())
    }
    fn get_recent(&self, count: usize) -> Result<Vec<EpisodicMemoryEntry>, Self::Error> {
        let e = self.entries.lock().unwrap();
        Ok(e.iter().rev().take(count).cloned().collect())
    }
    fn count(&self) -> Result<usize, Self::Error> {
        Ok(self.entries.lock().unwrap().len())
    }
    fn prune_older_than(
        &mut self,
        _cutoff: &chrono::DateTime<chrono::Utc>,
    ) -> Result<usize, Self::Error> {
        Ok(0)
    }
}

fn episodic_delta(source_device: &str, entry: &EpisodicMemoryEntry) -> SyncDelta {
    SyncDelta::new(
        "owner-1",
        source_device,
        "",
        SyncDomainKeys::MEMORY_EPISODIC,
        serde_json::to_vec(entry).unwrap(),
        1,
        SyncDeliveryMode::Guaranteed,
        None,
    )
}

// ── push side ─────────────────────────────────────────────────────────────────

#[test]
fn push_delta_fields_are_correct() {
    // Build a channel that shares its recorder so we can inspect the pushed delta
    // after the channel is moved into the service.
    let ch = FakeChannelHandle::default();
    let recorder = ch.clone_recorder();
    let svc = MemorySyncService::new(ch, VecEpisodicStore::default(), "dev-A");
    svc.push_memory_delta(
        "owner-9",
        SyncDomainKeys::GOALS,
        b"payload-bytes".to_vec(),
        SyncDeliveryMode::BestEffort,
    )
    .unwrap();

    let pushed = recorder.lock().unwrap();
    assert_eq!(pushed.len(), 1);
    let d = &pushed[0];
    assert_eq!(d.owner_id, "owner-9");
    assert_eq!(d.source_device_id, "dev-A");
    assert_eq!(d.target_device_id, ""); // broadcast
    assert_eq!(d.domain_key, SyncDomainKeys::GOALS);
    assert_eq!(d.payload, b"payload-bytes");
    assert!(d.is_broadcast());
    assert_eq!(d.delivery_mode, SyncDeliveryMode::BestEffort);
}

/// Channel variant that shares its recorder so a test can inspect pushes after
/// the channel is moved into the service.
#[derive(Default)]
struct FakeChannelHandle {
    pushed: std::sync::Arc<Mutex<Vec<SyncDelta>>>,
}
impl FakeChannelHandle {
    fn clone_recorder(&self) -> std::sync::Arc<Mutex<Vec<SyncDelta>>> {
        std::sync::Arc::clone(&self.pushed)
    }
}
impl ISyncChannel for FakeChannelHandle {
    type Error = Infallible;
    fn push_delta(&mut self, delta: &SyncDelta) -> Result<(), Self::Error> {
        self.pushed.lock().unwrap().push(delta.clone());
        Ok(())
    }
    fn receive_deltas(
        &self,
        _owner_id: &str,
    ) -> Result<Box<dyn Iterator<Item = Result<SyncDelta, Self::Error>>>, Self::Error> {
        Ok(Box::new(std::iter::empty()))
    }
    fn get_last_sequence(&self, _o: &str, _d: &str) -> Result<i64, Self::Error> {
        Ok(0)
    }
}

// ── receive side ──────────────────────────────────────────────────────────────

#[test]
fn receive_is_noop_until_started() {
    let entry = EpisodicMemoryEntry::new("hi", "hello");
    let inbound = vec![episodic_delta("dev-B", &entry)];
    let svc = MemorySyncService::new(
        FakeChannel::with_inbound(inbound),
        VecEpisodicStore::default(),
        "dev-A",
    );
    // Not started → drains nothing.
    assert_eq!(svc.receive_once("owner-1").unwrap(), 0);
    assert!(!svc.is_receiving());
}

#[test]
fn receive_applies_episodic_delta_from_other_device() {
    let entry = EpisodicMemoryEntry::new("what is the weather", "sunny");
    let inbound = vec![episodic_delta("dev-B", &entry)];
    let svc = MemorySyncService::new(
        FakeChannel::with_inbound(inbound),
        VecEpisodicStore::default(),
        "dev-A",
    );
    svc.start_receiving("owner-1").unwrap();
    assert!(svc.is_receiving());
    let applied = svc.receive_once("owner-1").unwrap();
    assert_eq!(applied, 1, "one episodic delta applied");
}

#[test]
fn receive_skips_own_echo() {
    let entry = EpisodicMemoryEntry::new("x", "y");
    // Delta sourced from the LOCAL device — must be skipped.
    let inbound = vec![episodic_delta("dev-A", &entry)];
    let svc = MemorySyncService::new(
        FakeChannel::with_inbound(inbound),
        VecEpisodicStore::default(),
        "dev-A",
    );
    svc.start_receiving("owner-1").unwrap();
    assert_eq!(svc.receive_once("owner-1").unwrap(), 0, "own echo skipped");
}

#[test]
fn receive_skips_other_domain_deltas() {
    let mut persona_delta = SyncDelta::new(
        "owner-1",
        "dev-B",
        "",
        SyncDomainKeys::PERSONA,
        b"not-an-episode".to_vec(),
        1,
        SyncDeliveryMode::Guaranteed,
        None,
    );
    // ensure created_at deterministic-ish (not asserted)
    persona_delta.sequence = 1;
    let svc = MemorySyncService::new(
        FakeChannel::with_inbound(vec![persona_delta]),
        VecEpisodicStore::default(),
        "dev-A",
    );
    svc.start_receiving("owner-1").unwrap();
    assert_eq!(svc.receive_once("owner-1").unwrap(), 0, "non-episodic domain ignored");
}

#[test]
fn receive_wrong_owner_is_noop() {
    let entry = EpisodicMemoryEntry::new("x", "y");
    let svc = MemorySyncService::new(
        FakeChannel::with_inbound(vec![episodic_delta("dev-B", &entry)]),
        VecEpisodicStore::default(),
        "dev-A",
    );
    svc.start_receiving("owner-1").unwrap();
    assert_eq!(svc.receive_once("someone-else").unwrap(), 0);
}

#[test]
fn stop_receiving_disables_drain() {
    let entry = EpisodicMemoryEntry::new("x", "y");
    let svc = MemorySyncService::new(
        FakeChannel::with_inbound(vec![episodic_delta("dev-B", &entry)]),
        VecEpisodicStore::default(),
        "dev-A",
    );
    svc.start_receiving("owner-1").unwrap();
    svc.stop_receiving().unwrap();
    assert!(!svc.is_receiving());
    assert_eq!(svc.receive_once("owner-1").unwrap(), 0);
}

// ── SyncReconciliation ────────────────────────────────────────────────────────

fn vv(pairs: &[(&str, i64)]) -> VersionVector {
    let mut m = HashMap::new();
    for (k, v) in pairs {
        m.insert(k.to_string(), *v);
    }
    VersionVector::new(m)
}

#[test]
fn version_vector_merge_takes_elementwise_max() {
    let a = vv(&[("n1", 5), ("n2", 2)]);
    let b = vv(&[("n2", 7), ("n3", 1)]);
    let m = SyncReconciliation::merge(&a, &b);
    assert_eq!(m.clocks.get("n1"), Some(&5));
    assert_eq!(m.clocks.get("n2"), Some(&7));
    assert_eq!(m.clocks.get("n3"), Some(&1));
}

#[test]
fn version_vector_dominance() {
    let a = vv(&[("n1", 5), ("n2", 3)]);
    let b = vv(&[("n1", 4), ("n2", 3)]);
    assert!(SyncReconciliation::a_dominates_b(&a, &b), "a strictly ahead on n1");
    assert!(!SyncReconciliation::a_dominates_b(&b, &a));
    // Equal vectors do not dominate (no strictly-greater component).
    assert!(!SyncReconciliation::a_dominates_b(&a, &a));
    // Concurrent vectors: neither dominates.
    let c = vv(&[("n1", 6), ("n2", 1)]);
    let d = vv(&[("n1", 2), ("n2", 9)]);
    assert!(!SyncReconciliation::a_dominates_b(&c, &d));
    assert!(!SyncReconciliation::a_dominates_b(&d, &c));
}

#[test]
fn last_writer_wins_prefers_later_then_ties_to_a() {
    let t1 = Utc.timestamp_opt(100, 0).unwrap();
    let t2 = Utc.timestamp_opt(200, 0).unwrap();
    let (w, v) = SyncReconciliation::last_writer_wins((t1, "old"), (t2, "new"));
    assert_eq!(w, t2);
    assert_eq!(v, "new");
    // Tie → a wins.
    let (w2, v2) = SyncReconciliation::last_writer_wins((t1, "A"), (t1, "B"));
    assert_eq!(w2, t1);
    assert_eq!(v2, "A");
}

// ── InMemoryGoalStore (sync IGoalStore) ───────────────────────────────────────

fn goal(id: &str, user: &str, status: GoalStatus) -> Goal {
    let mut g = Goal::new(id, user, "title", "desc", GoalPriority::Normal);
    g.status = status;
    g
}

#[test]
fn goal_store_upsert_get_list_delete() {
    let mut store = InMemoryGoalStore::new();
    store.upsert(goal("g1", "u1", GoalStatus::Active)).unwrap();
    store.upsert(goal("g2", "u1", GoalStatus::Completed)).unwrap();
    store.upsert(goal("g3", "u2", GoalStatus::Active)).unwrap();

    assert_eq!(store.get("g1").unwrap().unwrap().user_id, "u1");
    assert!(store.get("missing").unwrap().is_none());

    let u1 = store.list("u1").unwrap();
    assert_eq!(u1.len(), 2);
    // insertion order preserved
    assert_eq!(u1[0].id, "g1");
    assert_eq!(u1[1].id, "g2");

    let active = store.get_active("u1").unwrap();
    assert_eq!(active.len(), 1);
    assert_eq!(active[0].id, "g1");

    store.delete("g1").unwrap();
    assert!(store.get("g1").unwrap().is_none());
    assert_eq!(store.list("u1").unwrap().len(), 1);
}

#[test]
fn goal_store_upsert_replaces_in_place() {
    let mut store = InMemoryGoalStore::new();
    store.upsert(goal("g1", "u1", GoalStatus::Active)).unwrap();
    let mut updated = goal("g1", "u1", GoalStatus::Completed);
    updated.progress = 1.0;
    store.upsert(updated).unwrap();
    let got = store.get("g1").unwrap().unwrap();
    assert_eq!(got.status, GoalStatus::Completed);
    assert_eq!(got.progress, 1.0);
    // still a single entry, order intact
    assert_eq!(store.list("u1").unwrap().len(), 1);
}

#[test]
fn goal_store_rejects_blank_ids() {
    let mut store = InMemoryGoalStore::new();
    assert!(store.list("  ").is_err());
    assert!(store.get(" ").is_err());
    assert!(store.get_active("").is_err());
    assert!(store.delete("").is_err());
}
